### PR TITLE
Endpoint to get balances for a specified list of addresses

### PR DESCRIPTION
**Pull request by dimsavva**

### CODE REVIEW

**Comment by dangershony** [comment]
Ah great idea

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs [hunk 1]
```diff
@@ -79,48 +79,48 @@ public List<string> GetBlockIndexIndexes()
...
+         BsonElement inproc = currentOp.GetElement(0);
+         var arr = inproc.Value as BsonArray;
 
-            var ret = new List<IndexView>();
+         var ret = new List<IndexView>();
 
-            foreach (BsonValue bsonValue in arr)
+         foreach (BsonValue bsonValue in arr)
+         {
+            BsonElement? desc = bsonValue.AsBsonDocument?.GetElement("desc");
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:95]: I remember this was a bit of a specialized code handling to get it right, why change it? is this more generic and correct?

**Comment by dimsavva** [comment]
> @dimsavva this went bellow radar apparently Do we want this to go in?

Yeah, it is a useful feature I think instead of multiple api requests you can retrieve the balances of multiple addresses in one go, but up to you.

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs (hunk 1 above)
Comment (dimsavva) [reply src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:95]: From what I can see it's just null checks that I have added. Which lines of code are you referring to specifically?

File: src/Blockcore.Indexer.Core/Controllers/QueryController.cs [hunk 2]
```diff
@@ -39,6 +40,18 @@ public IActionResult GetAddress([MinLength(4)][MaxLength(100)] string address)
          return Ok(storage.AddressBalance(address));
       }
 
+      /// <summary>
+      /// Get the balance of multiple addresses.
+      /// </summary>
+      /// <param name="addresses"></param>
+      /// <returns></returns>
+      [HttpPost]
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Controllers/QueryController.cs:48]: I notice this is an http post, I guess we have to do that becuase we can't specify multiple addresses in the query string and get doesn't read from the body right?

It seems in this post that putting a body ion a get is actually fine now
https://stackoverflow.com/questions/978061/http-get-with-request-body

Thoughts?

File: src/Blockcore.Indexer.Core/Controllers/QueryController.cs [hunk 3]
```diff
@@ -39,6 +40,18 @@ public IActionResult GetAddress([MinLength(4)][MaxLength(100)] string address)
...
+      [Route("addressbalances")]
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Controllers/QueryController.cs:49]: This should follow the convention 
`address/balances` maybe? or 
"address/addressbalances"

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs [hunk 4]
```diff
@@ -680,6 +680,29 @@ public QueryAddress AddressBalance(string address)
...
+
+      /// <summary>
+      /// Calculates the balance for specified list of addresses.
+      /// </summary>
+      /// <param name="addresses"></param>
+      public List<AddressBalanceResult> MultipleAddressBalances(List<string> addresses)
+      {
+         var result = new ConcurrentBag<AddressBalanceResult>();
+
+         Parallel.ForEach(addresses, address =>
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:692]: We should run a distinct on the addresses list (just in case)

Comment (dangershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:692]: I wonder if parallel is really needed? some of this calls can take time if the address has many entries in history and has not been computed before

File: src/Blockcore.Indexer.Core/Models/AddressBalanceResult.cs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+namespace Blockcore.Indexer.Core.Models
+{
+   public class AddressBalanceResult
```
Comment (dangershony) [inline src/Blockcore.Indexer.Core/Models/AddressBalanceResult.cs:3]: Why not return a list of `QueryAddress` as already fetch that data from disk (apart for mempool)

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs (hunk 1 above)
Comment (dangershony) [reply src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:95]: ok yeah that's fine its just whitespaces

File: src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs [hunk 6]
```diff
@@ -680,6 +680,29 @@ public QueryAddress AddressBalance(string address)
...
+         {
+            result.Add(new AddressBalanceResult()
+            {
+               Address = address,
+               Balance = ComputeAddressBalance(address).Available
```
Comment (DavidGershony) [inline src/Blockcore.Indexer.Core/Storage/Mongo/MongoData.cs:697]: Isn't it better to sum the outputs in the UTXO table? can be much faster and easy to implement.

**Comment by dangershony** [comment]
We should probably try to get this pushed in the near future as this will be a good perf improvement for the wallet

**Comment by dangershony** [comment]
This will probably be replaced with the following PR
https://github.com/block-core/blockcore-indexer/pull/169

**Comment by dangershony** [comment]
Closing this as its gonna get replaced

**Pull request closed without merging**